src/dpp/events/guild_create.cpp

```diff
@@ -42,97 +42,105 @@ using namespace dpp;
  */
 void guild_create::handle(discord_client* client, json &j, const std::string &raw) {
 	json& d = j["d"];
-	bool newguild = false;
-	if (snowflake_not_null(&d, "id") == 0)
-		return;
-	dpp::guild* g = dpp::find_guild(snowflake_not_null(&d, "id"));
-	if (!g) {
-		g = new dpp::guild();
-		newguild = true;
-	}
-	g->fill_from_json(client, &d);
-	g->shard_id = client->shard_id;
-	if (!g->is_unavailable() && newguild) {
-		if (client->creator->cache_policy.role_policy != dpp::cp_none) {
-			/* Store guild roles */
-			g->roles.clear();
-			g->roles.reserve(d["roles"].size());
-			for (auto & role : d["roles"]) {
-				dpp::role *r = dpp::find_role(snowflake_not_null(&role, "id"));
-				if (!r) {
-					r = new dpp::role();
+	dpp::guild newguild;
+	dpp::guild* g = nullptr;
+	
+	if (client->creator->cache_policy.guild_policy == cp_none) {
+		newguild.fill_from_json(client, &d);
+		g = &newguild;
+	} else {
+		bool newguild = false;
+		if (snowflake_not_null(&d, "id") == 0)
+			return;
+		g = dpp::find_guild(snowflake_not_null(&d, "id"));
+		if (!g) {
+			g = new dpp::guild();
+			newguild = true;
+		}
+		g->fill_from_json(client, &d);
+		g->shard_id = client->shard_id;
+		if (!g->is_unavailable() && newguild) {
+			if (client->creator->cache_policy.role_policy != dpp::cp_none) {
+				/* Store guild roles */
+				g->roles.clear();
+				g->roles.reserve(d["roles"].size());
+				for (auto & role : d["roles"]) {
+					dpp::role *r = dpp::find_role(snowflake_not_null(&role, "id"));
+					if (!r) {
+						r = new dpp::role();
+					}
+					r->fill_from_json(g->id, &role);
+					dpp::get_role_cache()->store(r);
+					g->roles.push_back(r->id);
```

Comment:
as with role cache, without a guild cache we dont know what was removed to prune old values, over time the emoji cache would leak. This is because the update event just gives the current list, we have to compare against the existing cached list.